src/main/scala/sigmastate/lang/SigmaTyper.scala

```diff
@@ -339,17 +340,11 @@ class SigmaTyper {
           (newNode: (Value[T], Value[T]) => SValue): SValue = {
     val l1 = assignType(env, l).asValue[T]
     val r1 = assignType(env, r).asValue[T]
-    (l1.tpe, r1.tpe) match {
-      case (t1: SNumericType, t2: SNumericType) if t1 != t2 =>
-        val tmax = t1 max t2
-        val l = l1.upcastTo(tmax)
-        val r = r1.upcastTo(tmax)
-        newNode(l.asValue[T], r.asValue[T])
-      case (t1, t2) =>
-        if (t1 == t2)
-          newNode(l1, r1)
-        else
-          error(s"Invalid binary operation $op ($l1, $r1): type mismatch $t1 != $t2")
+    try {
+      newNode(l1, r1)
+    } catch {
+      case e: Throwable =>
+        throw new InvalidBinaryOperationParameters(s"operation $op: $e")
```

Comment:
@ergomorphic substring matching as error detection in sigma typer test seemed too fragile to me. How do you like the idea of specialized exceptions?

Reply:
@greenhat Yeah, that would be better. You can also think about encapsulating  some pretty printing logic there.
However, the missing piece of the puzzle is source text location info. 
We need to thread it somehow  from parser down to specializer, so that pretty printer would be able to show location in the original text.